theano/sandbox/gpuarray/opt.py

```diff
@@ -254,10 +255,43 @@ def local_gpu_elemwise(node):
     name = op.name
     if name:
         name = 'Gpu'+name
+
     res = GpuElemwise(scal_op, name=name,
-                      inplace_pattern=copy.copy(op.inplace_pattern),
-                      nfunc_spec=op.nfunc_spec)
-    return res
+                          inplace_pattern=copy.copy(op.inplace_pattern),
+                          nfunc_spec=op.nfunc_spec)
+
+    # If the elemwise operation is a pow, casts might be required on the
+    # inputs and or outputs because only the (float, float)->float and
+    # (double, double)->double cases are implemented at the moment.
+    if isinstance(op.scalar_op, Pow):
+        old_out_dtype = node.outputs[0].dtype
+        old_inp_dtypes = [inp.dtype for inp in node.inputs]
+        new_out_dtype = upcast("float32", *old_inp_dtypes)
```

Comment:
float16 is always supported for Elemwise operations on the GPU.  It might not be optimal, but it will work.  And since we cast the result back to the previous type, it won't make much of a difference.

However using floatX is fine for me.